docs/elements.md

```diff
@@ -311,19 +311,46 @@ This element supports additional preview options through [element extensions](el
+limit by using multiple `pl-file-upload` elements in the same question. To avoid unexpected errors or
+potentially misleading error messages for large file uploads, we recommend not using more than one
+`pl-file-upload` element per question.
+
 #### Sample element
 
 ![](elements/pl-file-upload.png)
 
 ```html
-<pl-file-upload file-names="foo.py, bar.c, filename with\, comma.txt"></pl-file-upload>
+<pl-file-upload
+  file-names="foo.py, bar.c, filename with\, comma.txt"
+  optional-file-names="foo.pdf, *.py, file_?.txt, file_[abc].txt, file with \\? as non-wildcard.pdf"
+></pl-file-upload>
 ```
 
 #### Customizations
 
-| Attribute    | Type     | Default | description                                                                                                                                    |
-| ------------ | -------- | ------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
-| `file-names` | CSV list | ""      | List of files that should and must be submitted. Commas in a filename should be escaped with a backslash, and filenames cannot contain quotes. |
+| Attribute             | Type     | Default | description                                                                                                                                                                                                                                                       |
+| --------------------- | -------- | ------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `file-names`          | CSV list | ""      | List of files that should and must be submitted. Commas in a filename should be escaped with a backslash, and filenames cannot contain quotes.                                                                                                                    |
+| `optional-file-names` | CSV list | ""      | List of files that can be submitted, but are optional. Wildcards are supported (see below). Commas should be escaped with a backslash, reserved wildcard characters in file names should be escaped with double backslashes, and filenames cannot contain quotes. |
+| `allow-blank`         | boolean  | false   | Whether or not an empty file input is allowed. By default, at least one file must be uploaded, even if only `optional-file-names` are specified                                                                                                                   |
+
+#### Supported wildcards
+
+Only `optional-file-names` supports a number of wildcards to allow a range of file names. Note that the use of wildcards enables multiple files to be uploaded for each wildcard pattern, all of which will be included in the submission.
```

Comment:
Given that only `optional-file-names` supports wildcards, the element wouldn't support something like "upload an arbitrary Python file implementing algorithm X and an arbitrary PDF file with a proof that your algorithm has runtime Y", right?

This comment shouldn't stand in the way of merging this, but I want to make sure it's something we've thought through. If that's not supported and we did want to consider supporting it in the future, an option would be to differentiate between "names" and "patterns" so that we can add required patterns in the future without worrying about backwards compatibility issues with `file-names`. That is, we could add `optional-file-patterns` in this PR and then `file-patterns` (or `required-file-patterns`) in a future one.

Reply:
I'm not a huge fan of ```optional-file-patterns``` as a name, especially since the parameter can also be used for just listing normal optional file names rather than patterns. Instead, I'd consider another option like ```enable-patterns``` that enables patterns for both ```file-names``` and ```optional-file-names```. The downside is that it has to be disabled by default for backwards compatibility, but on the other hand it also makes ```optional-file-names``` slightly easier to for less tech-savvy users since they don't have to worry about escaping anything if they just want a list of files.

Reply:
The idea of `enable-patterns` is interesting! I'd be ok with that, but I'd like to run this by a few others and build consensus. cc @jonatanschroeder and @eliotwrobson

Reply:
I like the idea of `enable-patterns`. This would also make it clearer when I list `*.pdf` if it's optional or if I require at least one PDF file, by using the pattern in `file-names` or `optional-file-names`. There would need to be a bit of juggling to ensure that at least one file for each pattern is present, but it is doable. It also removes the need for `allow-blank`.

An alternative is to have four options: `file-names`, `file-patterns`, `optional-file-names` and `optional-file-patterns`. Not sure which one is clearer, but the latter would likely simplify some of the implementation.

Reply:
> An alternative is to have four options: `file-names`, `file-patterns`, `optional-file-names` and `optional-file-patterns`. Not sure which one is clearer, but the latter would likely simplify some of the implementation.

I actually kindof like this idea? I think it's immediately obvious from reading the names what each attribute would do. Is the idea that `file-patterns` must be a pattern representing a finite list?

Reply:
Hmm the more I think about this the more awkward I find defining the exact semantics of ```file-patterns```. Another question is what happens if patterns overlap: Suppose we have "solution.*,*.py" - is a single file "solution.py" acceptable in this case? If not, how do we determine which pattern it gets assigned to? The logic for this seems increasingly unwieldy... 

I'm starting to wonder if the two options we currently have are actually sufficient (one very simple one, and one that is a catch-all that can be refined with a bit of backend code to potentially reject submissions that are nonsensical for a specific use case)...

Reply:
> Suppose we have "solution.*,*.py" - is a single file "solution.py" acceptable in this case? If not, how do we determine which pattern it gets assigned to? The logic for this seems increasingly unwieldy...

I think an instructor that leaves this kind of ambiguity would likely need to accept whatever the implementation does with that. And the implementation can treat this in whatever way is easier to manage. The current implementation even allows `file-names="solution.py,solution.py"`, with some weird side effects, but I don't think this has been a problem.

Reply:
Fair enough, I'll look into it then. Just to clarify though before I start: If you have a pattern like ```file-names="*.py"```, would that allow 1+ files, or should ```file-names``` enforce a 1:1 match of patterns and files. I personally think the latter is more useful since it allows cases like "upload [one] arbitrary Python file implementing algorithm X and [one] arbitrary PDF file with a proof that your algorithm has runtime Y" as proposed by @nwalters512 above.

Reply:
That is an interesting question. I agree with you that `file-patterns` could reasonably have a 1:1 match as default. Now the question is, should we allow for the pattern of allowing more than one?

Here's a proposal (brainstorming here):
* `file-patterns="*.pdf"` means one PDF file and only one.
* `optional-file-patterns="*.pdf"` means any number of PDF files, zero is fine. 
* `file-patterns="*.pdf" optional-file-patterns="*.pdf"` means at least one PDF file, but more are allowed.

Is this too complex? Would an instructor understand this? Would this implementation be reasonable? @nwalters512 thoughts?

Reply:
I've tried writing this down to see how complicated it works out, and I think it is reasonable (considering that this is all optional complexity that instructors aren't forced to care about). Note that the example in the very end might be too much for the actual example question and is more to show off how much flexibility this enables.

> Besides mandatory file names, <code>pl-file-upload</code> supports 3 other attributes to customize which files students can upload in a more fine-grained way:</p>
>   
>   <ul>
>     <li><code>optional-file-names</code> defines list of file names that students can (are not required to) upload.</li>
>     <li><code>file-patterns</code> defines list of file names that can include <a href="http://prairielearn.readthedocs.io/en/latest/elements/#pl-file-upload-element">wildcard patterns</a> and therefore allow a range of different names. Exactly one file must be provided to match each pattern.</li>
>     <li><code>optional-file-patterns</code> defines another list of <a href="http://prairielearn.readthedocs.io/en/latest/elements/#pl-file-upload-element">wildcard patterns</a> that can include wildcards. For each pattern, any number of files may be uploaded.</li></ul>
>A single <code>pl-file-upload</code> element can use any combination of mandatory and optional file names and patterns. For example, the following element requires the upload of a Python file named "solution.py" and at least 2 additional Python files, and also allows an optional "explanation.txt" file to be uploaded. 
```<pl-file-upload file-names="solution.py" file-patterns="*.py,*.py" optional-file-names="explanation.txt" optional-file-patterns="*.py"></pl-file-upload>```


From a more implementation-centric perspective, files would be mapped to the four lists in this way:
1. Any file name that is in `file-names` gets mapped to that name. All `file-names` must have a name assigned or the submission is invalid.
2. Any remaining file name that matches a `file-pattern` gets mapped to the first remaining matching pattern. All `file-patterns` must have a name assigned or the submission is invalid.
3. Any remaining file name that matches an `optional-file-name` gets mapped to that file name.
4. Any remaining file name that matches an `optional-file-pattern` gets added (no 1:1 mapping here).
5. Any remaining file name at this point is rejected.

I think although this involves a lot of steps internally, the ultimate behavior is not too complex (fill from required and most specific to optional and least specific) and supports most use cases I can think of.

Reply:
(I should also add that this has the added bonus of actually reducing the complexity of processing `optional-file-names` as it no longer has to separate names and patterns)

Reply:
I've updated the implementation, so I think all of @jonatanschroeder's examples above and mine should work as expected now.